### DISCUSSION TITLE
BACnet Services notification

### DISCUSSION THREAD

**Discussion by shashankgowdasd** [JoelBender/bacpypes#504] [Ideas]
Hi,
I am currently working on bacpypes(new to it), for implementing server model to my project and I have used the "minidevice.py" library sample code to implement the server and it is working fine and able to access the read and write services but I want to know
when the client request for the particular request(ex: readrequest) I need to know that property is being requested,So same for all the services I need to identify.

Regarding the above thing please suggest me which function I need to refer for the notification of request and to handle them properly @JoelBender @bbartling

**Comment by bbartling** [comment]
that's a great question : )

Am curious to know myself...i know BACnet is UDP and does UDP support that?

On a web app your looking for something like this right but the equivalent in a BACnet server like `mini_device.py`?
```
from flask import Flask, request

app = Flask(__name__)

@app.route("/")
def index():
    ip_address = request.remote_addr
    return "Your IP address is {}".format(ip_address)

if __name__ == "__main__":
    app.run(debug=True)
```

**Comment by JoelBender** [comment]
> I need to know [the] property [that] is being requested.

The `ReadWritePropertyServices` (defined in the `bacpypes.service.object` module) has a `do_ReadPropertyRequest()` method that is given the apdu which will be an instance of a `ReadPropertyRequest` (defined in the `bacpypes.apdu` module).
1. Take that class as a template and define your own, or inherit from it and just intercept that method.
2. Build your own stack like the `BIPSimpleApplication` and substitute your class in the parent class list.

> I know BACnet is UDP and does UDP support that?

All incoming PDUs from the socket layer have the source address which, for the Python socket library, is a tuple like `('1.2.3.4', 47808)`. This is converted into a BACpypes `Address` instance in the `pduSource` attribute.  Note that UDP is not Network Address Translation (NAT) friendly, so some intervening devices can muck around with sources and destinations.

**Comment by shashankgowdasd** [comment]
Thanks @bbartling @JoelBender for the suggestion..! 

it worked (ReadRequest,WriteRequest,ReadMultipleRequest) by inheriting the ReadWritePropertyServices  class and ReadWritePropertyMultipleServices and I got the call back notification printed by using my logic for but I want to know how to get the writerequest property values and readmultiplerequest property values during the request callback notification.

And also give me suggestion to implement the WriteMultipleRequest Service call back function..

**Reply by bbartling** [reply]
can you post your working code and `formatted nicely` for the eyes?

**Reply by shashankgowdasd** [reply]
Hi,
Please check and tell me to get the Writerequest Value and also give suggestion to implement the WriteMultipleproperty 

class CustomReadWritePropertyServices(ReadWritePropertyServices,ReadWritePropertyMultipleServices,WriteMultiplePropertyServices,WriteProperty):
    def do_WritePropertyRequest(self, apdu):
        try:
            objId = apdu.objectIdentifier
            Objprop = apdu.propertyIdentifier
            objprop_value = apdu.propertyArrayIndex

            if apdu.serviceChoice == 15:
                # Call the parent class's method to handle the WritePropertyRequest
                print("INFO : WriteRequest { objectIdentifier : ",objId ,",property =" ,Objprop,",property_value =",objprop_value," }")
                super().do_WritePropertyRequest(apdu)
        
        except PropertyError:
            raise ExecutionError(errorClass='property', errorCode='unknownProperty')

    def do_ReadPropertyRequest(self, apdu):
        try:
            ObjId = apdu.objectIdentifier
            Objprop = apdu.propertyIdentifier
            obj = self.get_object_id(ObjId)

            if apdu.serviceChoice == 12:
                result = obj.ReadProperty(apdu.propertyIdentifier, apdu.propertyArrayIndex)
                print("INFO : ReadRequest { objectIdentifier : ",ObjId ,",property =" ,Objprop,",property_value =",result, " }")
                # Call the parent class's method to handle the ReadPropertyRequest
                super().do_ReadPropertyRequest(apdu)

        except PropertyError:
            raise ExecutionError(errorClass='property', errorCode='unknownProperty')

**Reply by bbartling** [reply]
@shashankgowdasd would you ever be interested in building that into a bacpypes3 asyncio version? I am taking some notes on a bacpypes3 `mini_device_revisited.py`  (thanks @JoelBender for you help btw) and running it with Docker on rasp pi hardware:

https://gist.github.com/bbartling/573030a2facc97e178b2362606b9cdda

**Comment by JoelBender** [comment]
The APDU classes and their composite sequences all inherit from `DebugContents` which provides a `debug_contents()` method.  It's used by the debugging logging formatter, but you can call it directly.  To find out the contents of the more complicated requests and responses it's necessary to know the sub-structures (lists of things like SequenceOfX), and their composite elements (the named pieces of X) and these are dumped out recursively by `debug_contents()`.  For `WritePropertyMultipleRequest`, use the force and read the source (or the Standard!) and see it is made up of a sequence of (a.k.a. list or array) of `WriteAccessSpecification` which is in turn composed of an object identifier and a list of property values, etc.

If you want to analyze the traffic in some other application, with BACpypes3 you can turn these objects into JSON blobs like this:
```python
>>> from bacpypes3.apdu import WhoIsRequest
>>> from bacpypes3.json import sequence_to_json
>>> 
>>> who_is = WhoIsRequest(deviceInstanceRangeLowLimit=0, deviceInstanceRangeHighLimit=99)
>>> 
>>> sequence_to_json(who_is)
{'device-instance-range-low-limit': 0, 'device-instance-range-high-limit': 99}
```
and you get the same content without the BACnet encoding/decoding rules and matching the ASN.1 notation in the standard.  Publish this to an MQTT broker and you've got yourself a nice little stream that's easy to parse.  In fact, if you do this at the application layer rather than deep in specific services, you'll see all kinds of behavior.

This kind of thing is built into the Wireshark family, try running `tshark -l -n -T json` and you'll get all kinds of good stuff:
```json
        "bacapp": {
          "bacapp.type": "1",
          "bacapp.unconfirmed_service": "8",
          "bacapp.who_is.low_limit": "0",
          "bacapp.who_is.low_limit_tree": {
            "Context Tag: 0, Length/Value/Type: 1": {
              "bacapp.tag_class": "1",
              "bacapp.context_tag_number": "0",
              "bacapp.LVT": "1"
            }
          },
          "bacapp.who_is.high_limit": "99",
          "bacapp.who_is.high_limit_tree": {
            "Context Tag: 1, Length/Value/Type: 1": {
              "bacapp.tag_class": "1",
              "bacapp.context_tag_number": "1",
              "bacapp.LVT": "1"
            }
          }
        }
```

**Reply by shashankgowdasd** [reply]
Hi,
Can you suggest me how to access the "whohasproperty" request ,because I have tried by inheriting the Discover.py , Sample3_WhoHasIHaveApplication.py requests and checked but no response and I am not able to detect the whohas - Ihave services on wireshark too

**Comment by JoelBender** [comment]
You run the COV server sample application and turn on debugging for the service:
```shell
$ python3 samples/cov-server.py --debug bacpypes3.service.device.WhoHasIHaveServices
```
Then from another device generate a Who-Has request.  You see a few lines like:
```python
DEBUG:bacpypes3.service.device.WhoHasIHaveServices:do_WhoHasRequest <bacpypes3.apdu.WhoHasRequest...>
...
DEBUG:bacpypes3.service.device.WhoHasIHaveServices:i_have (<ObjectType: analog-value>, 1) 'Wowzers!' address=<IPv4Address ...>
```
The first line is where the application is receiving the request and the `i_have()` method call is it sending the response.

To generate a request from some other device running the BACpypes3 module:
```
$ python3 -m bacpypes3
> whohas analog-value,1
999 analog-value,1 'Wowzers!'
```
